File: backend/app/schemas/orchestration.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


# ─── Step ─────────────────────────────────────────────────────────────────

class OrchestrationStepBase(BaseModel):
    workflow_id: str
    order_index: int = 0
    retry_count: int = Field(default=0, ge=0, le=10)
    retry_delay_seconds: int = Field(default=30, ge=0, le=3600)
    timeout_seconds: int = Field(default=0, ge=0, le=86400)
    on_failure: str = Field(default="stop", pattern="^(stop|continue)$")


class OrchestrationStepCreate(OrchestrationStepBase):
    pass


class OrchestrationStepUpdate(BaseModel):
    workflow_id: Optional[str] = None
    order_index: Optional[int] = None
    retry_count: Optional[int] = Field(default=None, ge=0, le=10)
    retry_delay_seconds: Optional[int] = Field(default=None, ge=0, le=3600)
    timeout_seconds: Optional[int] = Field(default=None, ge=0, le=86400)
    on_failure: Optional[str] = None


class OrchestrationStepResponse(OrchestrationStepBase):
    id: str
    orchestration_id: str
    workflow_name: Optional[str] = None

    model_config = {"from_attributes": True}


# ─── Orchestration ─────────────────────────────────────────────────────────

class OrchestrationCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    cron_expression: str = Field(..., min_length=3)
    is_active: bool = True
    on_error: str = Field(default="stop", pattern="^(stop|continue)$")
    steps: list[OrchestrationStepCreate] = Field(default_factory=list)


class OrchestrationUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = None
    cron_expression: Optional[str] = None
    is_active: Optional[bool] = None
    on_error: Optional[str] = None
    steps: Optional[list[OrchestrationStepCreate]] = None  # Tüm adımları değiştirir


class OrchestrationResponse(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    cron_expression: str
    is_active: bool
    on_error: str
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    steps: list[OrchestrationStepResponse] = []

    model_config = {"from_attributes": True}


# ─── Run result ────────────────────────────────────────────────────────────

class OrchestrationRunResult(BaseModel):
    orchestration_id: str
    orchestration_name: str
    total_steps: int
    completed_steps: int
    failed_steps: int
    skipped_steps: int
    execution_ids: list[str]
    status: str  # success | partial | failed

File: frontend/src/types/orchestration.ts
export interface OrchestrationStep {
  id: string
  orchestration_id: string
  workflow_id: string
  workflow_name?: string
  order_index: number
  retry_count: number
  retry_delay_seconds: number
  timeout_seconds: number
  on_failure: 'stop' | 'continue'
}

export interface Orchestration {
  id: string
  name: string
  description?: string
  cron_expression: string
  is_active: boolean
  on_error: 'stop' | 'continue'
  last_run_at?: string
  next_run_at?: string
  created_at: string
  updated_at: string
  steps: OrchestrationStep[]
}

export interface OrchestrationStepCreate {
  workflow_id: string
  order_index: number
  retry_count: number
  retry_delay_seconds: number
  timeout_seconds: number
  on_failure: 'stop' | 'continue'
}

export interface OrchestrationCreate {
  name: string
  description?: string
  cron_expression: string
  is_active: boolean
  on_error: 'stop' | 'continue'
  steps: OrchestrationStepCreate[]
}

export interface OrchestrationUpdate {
  name?: string
  description?: string
  cron_expression?: string
  is_active?: boolean
  on_error?: 'stop' | 'continue'
  steps?: OrchestrationStepCreate[]
}

export interface OrchestrationRunResult {
  orchestration_id: string
  orchestration_name: string
  total_steps: number
  completed_steps: number
  failed_steps: number
  skipped_steps: number
  execution_ids: string[]
  status: 'success' | 'partial' | 'failed'
}